list-sync/src/diff.ts

```diff
@@ -0,0 +1,487 @@
+import { Contract, JsonRpcProvider, Log } from 'ethers';
+import { existsSync, readFileSync } from 'fs';
+import { dirname, join } from 'path';
+import { loadAllAllowlists, AllowlistNormalized, SelectorState } from './allowlist.js';
+import { parse } from 'yaml';
+
+const TRANSFER_SELECTOR = '0xa9059cbb';
+const APPROVE_SELECTOR = '0x095ea7b3';
+const ZERO_SELECTOR = '0x00000000';
+const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';
+
+interface ChainConfig {
+  slug: string;
+  chainId: number;
+  rpcEnv: string;
+  guard: string;
+  module: string;
+  tags: string[];
+  startBlock: number;
+  adminSafe: string;
+}
+
+interface ChainsFile {
+  chains: ChainConfig[];
+}
+
+interface GuardDiffEntry {
+  target: string;
+  selector: string;
+  desired?: SelectorState;
+  actual?: SelectorState;
+  signature?: string;
+}
+
+interface ModuleDiffEntry {
+  target: string;
+  selector: string;
+  signature?: string;
+}
+
+export interface ChainDiffResult {
+  chain: ChainConfig;
+  guardMissing: GuardDiffEntry[];
+  guardMismatched: GuardDiffEntry[];
+  guardExtra: GuardDiffEntry[];
+  moduleMissing: ModuleDiffEntry[];
+  moduleExtra: ModuleDiffEntry[];
+}
+
+function normalizeAddress(address: string, context: string): string {
+  if (!address.startsWith('0x') || address.length !== 42) {
+    throw new Error(`Invalid address ${address} (${context})`);
+  }
+  return address.toLowerCase();
+}
+
+function normalizeSelector(selector: string, context: string): { selector: string; signature?: string } {
+  if (!/^0x[0-9a-fA-F]{8}$/.test(selector)) {
+    throw new Error(`Invalid selector ${selector} (${context})`);
+  }
+  return { selector: selector.toLowerCase() };
+}
```

Comment:
normalizeAddress and normalizeSelector defined in both diff and allowlist. Maybe intentional for different contexts - reading from chain vs addressing user input.